---
title: "Trajectories"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal

- To give you some experience with methodology focused on interpreting changes in cells as trajectories.
- To introduce you to the Monocle R library.

## Introduction


################
## Load the data
################

There are many ways to load data into monocle, we will use one way here (reading in a text file). Monocle is interoperable with other packages like Seurat and Scater and has functionality to directly pull in SCESets objects into CellDataSet (native Monocle objects). We will need to load three types of data.

**Expression File**:  The expression data should have genes in the rows and cells in the columns of the martix. The measurements can be relative data (eg. TPM units) or absolution counts (if using UMIs). Monocle should be given either counts that are corrected using UMIs or relative measurements that are controlled for transcript length like TPM. Monocle will work directly with CellRanger expression matrix outputs.

**Sample Data**: This phenotypic data should have cells as rows and cell metadata in the columns of the matrix. These metadata can be many types of measurements including QC like complexity, metadata like cell type or experimental cohorts, or assay details like day captured or kit lots. The number of rows (cells) should match the number of columns (cells) in the expression matrix and should have the same names.

**Gene Data**: These gene or transcript descriptions should have genes (or transcripts) as rows and columns as gene metadata like biotype or GC content. The number of rows (genes) should be the same in this file and the expression file; row names between these two files should also match. One of the columns of this file should be named "gene_short_name" and contain short names for the genes. These names will be used in the code when making the CellDataSet object.

```{r}
# Demo Files
expression_data <- "TODO_TODO"
sample_data <- "TODO_TODO"
gene_data <- "TODO_TODO"

# Read in files
## Here we are reading in the expression matrix as a sparse matrix.
## Monocle will continue using sparse matrices if given a sparse matrix.
expression <- as( read.table(expression_data), sparseMatrix )
samples <- read.delim(sample_data)
genes <- read.delim(gene_data)
```

Monocle needs to understand what is the distribution of your data before it begins. There are recommended settings depending on if you are using TPMs (Tobit; tobit()), log-normal TPM (Gaussian; gaussianff()), or UMI corrected counts (Negative binomial; negativebinomial.size(). Here we use TODOTODO . Let's look at your data before working with it to confirm the distribution that best models your data.

```{r}

# Here we randomly sample from the data to see the distribution.
## We use a seed here for reproducibility, this is not recommended in analysis outside of a tutorial.
# TODO
# Randomly sample from the data and plot a histogram.
# use a seed
```

The sampling confirms that we should select the negative binomial distribution. Let's use that when we make the CellDataSet object. If this is mispecified you can get poor performance. Feel free to play with this after completing this tutorial. Other options can be discovered in the documentation by typing "? newCellDataSet".

```{r}
# Create a CellDataSet object
sample_annot <- new("AnnotatedDataFrame", data = samples)
gene_annot <- new("AnnotatedDataFrame", data = genes)
mon_data <- newCellDataSet( as.matrix(expression), phenoData = sample_annot, featureData = gene_annot, expressionFamily = negativebinomial.size())
```

Because we used the negative binomial distribution, we also need to calculate size factors and dispersions that are respectively used internally to normalize measurements between cells and perform differential expression using the negative binomial distribution.

```{r}
# Calculate size factors for internal normalization
## Show the object before and after so they can see what the function does. TODO
mon_data <- estimatedSizeFactors(mon_data)
# Calculate dispersion for differential expression
## Show the object before and after so they can see what the function does. TODO
mon_data <- estimateDispersions(mon_data)
```

###############
## Filter the data
###############

We have seen in previous tutorials that filtering low-quality cells is important. If not already performed, it is important we filter the cells. Monocle filters cells beased on 

```{r}
mon_data <- detectGenes(mon_data, min_expr = 0.1)
print(head(fData(mon_data)))
qc_genes <- row.names(subset(fData(mon_data), num_cells_expressed >= 10))
```

This data set has alread been filtered in the following way, You do not need to run this code, but let's talk through it.
```{r}
### No need to run this code
valid_cells <- row.names(subset(pData(mon_data),
    # Keep cells with a value 1 for cells in well
    Cells.in.Well == 1 &
    # Keep controls
    Control == FALSE &
    # Keep cells with clump metadata false
    Clump == FALSE &
    # Keep cells with debris metadata false
    Debris == FALSE &
    # Keep cells with more than 1000000 mapped fragments
    Mapped.Fragments > 1000000))
### No need to run this code
# This would perform the cell filtering
mon_data <- mon_data[,valid_cells]
### No need to run this code
```

Let's look at the ditribution of mRNA totals across the cells.

```{r}
pData(HSMM)$Total_mRNAs <- Matrix::colSums(exprs(HSMM))

HSMM <- HSMM[,pData(HSMM)$Total_mRNAs < 1e6]

upper_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) +
            2*sd(log10(pData(HSMM)$Total_mRNAs)))
lower_bound <- 10^(mean(log10(pData(HSMM)$Total_mRNAs)) -
            2*sd(log10(pData(HSMM)$Total_mRNAs)))

qplot(Total_mRNAs, data = pData(HSMM), color = Hours, geom =
"density") +
geom_vline(xintercept = lower_bound) +
geom_vline(xintercept = upper_bound)
```

```{r}
HSMM <- HSMM[,pData(HSMM)$Total_mRNAs > lower_bound &
      pData(HSMM)$Total_mRNAs < upper_bound]
HSMM <- detectGenes(HSMM, min_expr = 0.1)
```

```{r}
# Log-transform each value in the expression matrix.
L <- log(exprs(HSMM[expressed_genes,]))

# Standardize each gene, so that they are all on the same scale,
# Then melt the data with plyr so we can plot it easily
melted_dens_df <- melt(Matrix::t(scale(Matrix::t(L))))

# Plot the distribution of the standardized gene expression values.
qplot(value, geom = "density", data = melted_dens_df) +
stat_function(fun = dnorm, size = 0.5, color = 'red') +
xlab("Standardized log(FPKM)") +
ylab("Density")
```

###############
## Classify cells
##############

```{r}
MYF5_id <- row.names(subset(fData(mon_data), gene_short_name == "MYF5"))
ANPEP_id <- row.names(subset(fData(mon_data),
    gene_short_name == "ANPEP"))

cth <- newCellTypeHierarchy()
cth <- addCellType(cth, "Myoblast", classify_func =
    function(x) { x[MYF5_id,] >= 1 })
cth <- addCellType(cth, "Fibroblast", classify_func = function(x)
 { x[MYF5_id,] < 1 & x[ANPEP_id,] > 1 })
mon_data <- classifyCells(mon_data, cth, 0.1)
table(pData(mon_data)$CellType)
```

Monocle hs support for various types of clustering. This was already covered in other tutorials but feel free to explore this if you complete this tutorial with time. The reference to the full tutorial is found below.

############
## Constructing trajectories
############

Choose genes to use to define the trajectory.

```{r}
diff_test_res <- differentialGeneTest(HSMM_myo[expressed_genes,],
              fullModelFormulaStr = "~Media")
ordering_genes <- row.names (subset(diff_test_res, qval < 0.01))
HSMM_myo <- setOrderingFilter(HSMM_myo, ordering_genes)
plot_ordering_genes(HSMM_myo)
```

Reduce the dimensionality of the data.

```{r}
HSMM_myo <- reduceDimension(HSMM_myo, max_components = 2,
    method = 'DDRTree')
```

Order cells

```{r}
HSMM_myo <- orderCells(HSMM_myo)
plot_cell_trajectory(HSMM_myo, color_by = "Hours")
plot_cell_trajectory(HSMM_myo, color_by = "State")
```

View a specific time as a branch.

```{r}
plot_cell_trajectory(HSMM_myo, color_by = "State") +
    facet_wrap(~State, nrow = 1)
```

View gene within the context of the states.

```{r}
blast_genes <- row.names(subset(fData(HSMM_myo),
gene_short_name %in% c("CCNB2", "MYOD1", "MYOG")))
plot_genes_jitter(HSMM_myo[blast_genes,],
    grouping = "State",
    min_expr = 0.1)
```

View genes across pseudotime.
```{r}
HSMM_expressed_genes <-  row.names(subset(fData(HSMM_myo),
num_cells_expressed >= 10))
HSMM_filtered <- HSMM_myo[HSMM_expressed_genes,]
my_genes <- row.names(subset(fData(HSMM_filtered),
          gene_short_name %in% c("CDK1", "MEF2C", "MYH3")))
cds_subset <- HSMM_filtered[my_genes,]
plot_genes_in_pseudotime(cds_subset, color_by = "Hours")
```

Monocle supplies many ways to perform differential expression, of interest is performing finding genes that change as a function of the psuedotime.

```{r}
to_be_tested <- row.names(subset(fData(HSMM),
gene_short_name %in% c("MYH3", "MEF2C", "CCNB2", "TNNT1")))
cds_subset <- HSMM_myo[to_be_tested,]

diff_test_res <- differentialGeneTest(cds_subset,
fullModelFormulaStr = "~sm.ns(Pseudotime)")

diff_test_res[,c("gene_short_name", "pval", "qval")]

plot_genes_in_pseudotime(cds_subset, color_by = "Hours")
```

Clustering by pseudotemporal expression

```{r}
diff_test_res <- differentialGeneTest(HSMM_myo[marker_genes,],
              fullModelFormulaStr = "~sm.ns(Pseudotime)")
sig_gene_names <- row.names(subset(diff_test_res, qval < 0.1))
plot_pseudotime_heatmap(HSMM_myo[sig_gene_names,],
                num_clusters = 3,
                cores = 1,
                show_rownames = T)


```

Multi-factorial differental expression analysis

```{r}
to_be_tested <-
    row.names(subset(fData(HSMM),
        gene_short_name %in% c("TPM1", "MYH3", "CCNB2", "GAPDH")))

cds_subset <- HSMM[to_be_tested,]

diff_test_res <- differentialGeneTest(cds_subset,
                        fullModelFormulaStr = "~CellType + Hours",
                        reducedModelFormulaStr = "~Hours")
diff_test_res[,c("gene_short_name", "pval", "qval")]
plot_genes_jitter(cds_subset,
    grouping = "Hours", color_by = "CellType", plot_trend = TRUE) +
        facet_wrap( ~ feature_label, scales= "free_y")
```

Analyzing branches in singel-cell trajectories

```{r}
lung <- load_lung()
plot_cell_trajectory(lung, color_by = "Time")

BEAM_res <- BEAM(lung, branch_point = 1, cores = 1)
BEAM_res <- BEAM_res[order(BEAM_res$qval),]
BEAM_res <- BEAM_res[,c("gene_short_name", "pval", "qval")]

plot_genes_branched_heatmap(lung[row.names(subset(BEAM_res,
                                          qval < 1e-4)),],
                                          branch_point = 1,
                                          num_clusters = 4,
                                          cores = 1,
                                          use_gene_short_name = T,
                                          show_rownames = T)

lung_genes <- row.names(subset(fData(lung),
          gene_short_name %in% c("Ccnd2", "Sftpb", "Pdpn")))
plot_genes_branched_pseudotime(lung[lung_genes,],
                       branch_point = 1,
                       color_by = "Time",
                       ncol = 1)
```

## Sources

This tutorial is ***strongly*** influenced by Monocle's Tutorials, and so we give them the credit for this tutorial. Explore Monocle here: http://cole-trapnell-lab.github.io/monocle-release/tutorials
